### PR TITLE
Adds fist draft of post on telemetry

### PR DESCRIPTION
**Pull request by iamkirkbater**
Adds the first draft of a post on telemetry in go

### CODE REVIEW

File: content/posts/go-telemetry.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+---
+title: "Telemetry in go - Readers"
+date: 2023-07-24T20:32:00-05:00
+draft: true
+---
+
+I was struggling to understand when to use the two main types of readers that come within the OpenTelemetry go package, and wanted to share what I've learned. This is the intro post to a series where I'm going to cover some telemetry things in go, ending with creating a plug-and-play telemetry add-on for the cobra-cli.
```
Comment (geowa4) [inline content/posts/go-telemetry.md]: Capitalize "Go"; link to https://cobra.dev/.

File: content/posts/go-telemetry.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+# What is a Reader
+
+So there are two types of "Readers" that I'm going to be talking about. A Reader in OpenTelemetry is essentially something that runs locally within your application that aggregates the given metrics your application is programmed to output.  The Reader's job is only to aggregate them, not to send them off, which is the job of the Exporter. We can see this with the [Reader interface](https://github.com/open-telemetry/opentelemetry-go/blob/64e76f8be45c9f4df85344249fad0fb72cff1230/sdk/metric/reader.go#L54) where we see that the only public methods are RegisterProducer, Collect, ForceFlush, and Shutdown. You can read about what the various methods should do in the code comments, the open-telemetry maintainers have done a very good job of adding them.
```
Comment (pellared) [inline content/posts/go-telemetry.md:15]: Side note: In the next release, the `PeriodicReader` and `ManualReader` types are exported so that the documented in pkg.dev.go. I hope it will be more approachable for the users.

File: content/posts/go-telemetry.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+So what are the two types of Readers that we're going to focus on? Specifically, I want to dive into the difference between a PeriodicReader and a ManualReader. A PeriodicReader, as the name suggests, performs a read (calls the Collect function) at a provided interval. With a Manual Reader the Collect function needs to be manually called.
+
+## Periodic Reader
+
+Most of the code examples I was able to find during my research all used the PeriodicReader implementation, which is probably what you want to do for any given long-lived server-side application. It makes sense to read and then pass your metrics to your exporter at a given interval while your application hums along in those cases.
```
Comment (pellared) [inline content/posts/go-telemetry.md:21]: I do not think it is true.

1. It is fine to use periodic reader even is short-living apps as `Shutdown` flushes pending telemetry.
2. Manual readers are used for long-living apps as well. The pull-based exporters (like Prometheus) take use of a manual reader.

File: content/posts/go-telemetry.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+However, in my case, I don't have a long-lived application. Some of the commands in the CLI that I'm attempting to instrument run for sub 1s, so having a PeriodicReader implementation that collects every 1s doesn't really make sense in my use case. However, it's worthy to note that it's not exactly a problem that the PeriodicReader doesn't hit it's interval, I noticed in my tests that it always still seemed to get the metrics from those short-lived runs. My ed-chu-ma-kate'd guess is that the Exporter I was using - when it receives a Shutdown command from a `defer`'d function at the end of the command invocation - will run a last `Collect` and export the metrics before gracefully exiting.
```
Comment (pellared) [inline content/posts/go-telemetry.md]: > My ed-chu-ma-kate'd guess is that the Exporter I was using - when it receives a Shutdown command from a `defer`'d function at the end of the command invocation - will run a last `Collect` and export the metrics before gracefully exiting.

Yup. See https://github.com/open-telemetry/opentelemetry-go/blob/088ac8e179bd30ee39c81278010f8d3b45ba45be/sdk/metric/periodic_reader.go#L320

File: content/posts/go-telemetry.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+While this _works_ it just _feels_ wrong to rely on. What if my command runs for _more_ than 1s and it `Collects` and then exports that multiple times? I mean, it's probably not a _real_ problem because I'd assume that either the Reader would have dumped the metric and wouldn't send it again, or that the server I'm sending these metrics to would be able to intelligently dedupe any duplicate metrics - but it still just doesn't sit right with me to use something only because it accidentally works for your purpose rather than using it for its intended purpose.
```
Comment (pellared) [inline content/posts/go-telemetry.md]: Multiple exports gives a lot of value in observability. Otherwise, you have no information until the command is finished. How would you know that e.g. when a command is stuck because you have some runtime problems?

Take notice you can also call `ForceFlush` on the periodic reader to make an on-demand collect+export.

File: content/posts/go-telemetry.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+At least, that's how I first understood it. Looking back on it now, those unit tests show _exactly_ how to use the Manual Reader - the piece that wasn't clicking for me was that the unit tests show that the expected output is an [empty error with no metrics](https://github.com/open-telemetry/opentelemetry-go/blob/64e76f8be45c9f4df85344249fad0fb72cff1230/sdk/metric/manual_reader_test.go#L94). What I was missing here is that while in this unit test we're expecting an empty error AND an empty struct, its OUR job as the consumer to pass the empty struct pointer to the `Collect` function to be filled, or more succinctly - when they call it a Manual Reader - it's _entirely_ manual. Where I thought this empty struct was just being compared in the test, it's actually what we use to populate our metric data, then we can pass that to our exporter.
```
Comment (pellared) [inline content/posts/go-telemetry.md]: This is valuable feedback 👍 I made a little improvement: https://github.com/open-telemetry/opentelemetry-go/pull/4361

File: content/posts/go-telemetry.md (hunk 3 above)
Comment (iamkirkbater) [reply content/posts/go-telemetry.md:21]: Thanks for this! While I don't correct this paragraph directly, I've added another paragraph that compares the two and implements this feedback directly.

File: content/posts/go-telemetry.md (hunk 5 above)
Comment (iamkirkbater) [reply content/posts/go-telemetry.md]: This makes a ton of sense. I've included this feedback!
